src/mjolnir/ingest_transit.cc

```diff
@@ -0,0 +1,778 @@
+
+  auto path = pt.get<std::string>("mjolnir.transit_feeds_dir");
+
+  std::set<GraphId> tiles;
+  const auto& tile_level = TileHierarchy::levels().back();
+  std::unordered_map<GraphId, tileTransitInfo> tile_map;
+
+  filesystem::recursive_directory_iterator gtfs_feed_itr(path);
+  filesystem::recursive_directory_iterator end_file_itr;
+  for (; gtfs_feed_itr != end_file_itr; ++gtfs_feed_itr) {
+    if (filesystem::is_directory(gtfs_feed_itr->path())) {
+      auto feed_path = gtfs_feed_itr->path().string();
+      gtfs::Feed feed(feed_path);
+      feed.read_feed();
+      const auto& stops = feed.get_stops();
+
+      for (const auto& stop : stops) {
+        gtfs::Id currStopId = stop.stop_id;
+        float x = stop.stop_lon;
+        float y = stop.stop_lat;
+
+        // add unique GraphId and stop
+        // might have to typecast these doubles into uint32_t
+        GraphId currId = GraphId(tile_level.tiles.TileId(y, x), 3, 0);
+
+        tileTransitInfo currTile;
+        if (tile_map.find(currId) != tile_map.end()) {
+          currTile = tile_map[currId];
+        } else {
+          currTile.graphId = currId;
+        }
+        currTile.tile_stops.insert({currStopId, feed_path});
+        gtfs::StopTimes currStopTimes = feed.get_stop_times_for_stop(currStopId);
+
+        for (const auto& stopTime : currStopTimes) {
+          // add trip, route, agency and service_id from stop_time
+          // could use std::set to not check if it already exists
+
+          // use a set instead to keep track of these points
+          gtfs::Trip currTrip = *(feed.get_trip(stopTime.trip_id));
+          currTile.tile_services.insert({currTrip.service_id, feed_path});
+          currTile.tile_trips.insert({currTrip.trip_id, feed_path});
+
+          currTile.tile_shapes.insert({currTrip.shape_id, feed_path});
+          // remember the shape id such that we can access it later
+          gtfs::Route currRoute = *(feed.get_route(currTrip.route_id));
+          currTile.tile_routes.insert({currRoute.route_id, feed_path});
+          currTile.tile_agencies.insert({currRoute.agency_id, feed_path});
+          // add to the queue of tiles
+        }
+        tile_map[currId] = currTile;
+      }
+    }
+  }
+  std::priority_queue<tileTransitInfo> prioritized;
+  for (auto it = tile_map.begin(); it != tile_map.end(); it++) {
+    prioritized.push(it->second);
+  }
+  return prioritized;
+}
+
+std::unordered_map<gtfs::Id, GraphId> write_stops(Transit& tile, const tileTransitInfo& tile_info) {
+
+  const auto& tile_stopIds = tile_info.tile_stops;
+  auto node_id = tile_info.graphId;
+  feedCache stopFeeds;
+
+  std::unordered_map<gtfs::Id, GraphId> stop_graphIds;
+  // loop through all stops inside the tile
+  for (const feedObject& feed_stop : tile_stopIds) {
+    const auto& tile_stopId = feed_stop.id;
+    const auto& feed = stopFeeds(feed_stop);
+    auto* node = tile.add_nodes();
+    gtfs::Stop tile_stop = *(feed.get_stop(tile_stopId));
```

Comment:
there's no way this doesn't return a valid stop? think I'd still protect for it, not too much experience with GTFS real data but I'm sure they can be very messed up.

Reply:
In `select_transit_tiles` I loop through the stop data from the feed, so all the stops should exist, but just in case the feed doesn't include stop_id, I just won't process the stop if it doesn't have an id.

Reply:
yeah i do think its impossible for this to get screwed up. the feedObject is populated by looping over all the stops in the feed as chris is saying. this is just going back on a second pass and pu lling out the rest of the info for the stop (once we know all the stops that will be in a tile)